### PR TITLE
Helper for https://github.com/CEED/libCEED/pull/900

### PR DESCRIPTION
**Pull request by rezgarshakeri**

### CODE REVIEW

**Comment by rezgarshakeri** [comment]
@jedbrown @jeremylt , I just created this branch to debug https://github.com/CEED/libCEED/pull/900. 
I only added `BasisCreateFromPlex` that means `basis_x` and `basis_u` are created from that function.
All the test passed, but I don't get convergence for hex (unfortunately I didn't check convergence rate in the open PR we have, I checked the test only).
For example `./bps -problem bp2 -degree 3 -local_nodes 500 -ksp_max_it 100`  (first issue: need more ksp iteration unlike main) and the errors are
`local_nodes 500`
```
  Mesh:
    Number of 1D Basis Nodes (P)       : 4
    Number of 1D Quadrature Points (Q) : 5
    Global nodes                       : 700
    Local Elements                     : 18
    Owned nodes                        : 700
    DoF per node                       : 3
  KSP:
    KSP Type                           : cg
    KSP Convergence                    : CONVERGED_RTOL
    Total KSP Iterations               : 13
    Final rnorm                        : 9.306932e-12
  Performance:
    Pointwise Error (max)              : 7.320278e-11
    CG Solve Time                      : 0.00162257 (0.00162257) sec
    DoFs/Sec in CG                     : 16.8251 (16.8251) million
```

`local_nodes 1000`
```
 Mesh:
    Number of 1D Basis Nodes (P)       : 4
    Number of 1D Quadrature Points (Q) : 5
    Global nodes                       : 1456
    Local Elements                     : 40
    Owned nodes                        : 1456
    DoF per node                       : 3
  KSP:
    KSP Type                           : cg
    KSP Convergence                    : CONVERGED_RTOL
    Total KSP Iterations               : 31
    Final rnorm                        : 3.623469e-10
  Performance:
    Pointwise Error (max)              : 2.201151e-08
    CG Solve Time                      : 0.00509974 (0.00509974) sec
    DoFs/Sec in CG                     : 26.552 (26.552) million
```

`local_nodes 2000`
```
  Mesh:
    Number of 1D Basis Nodes (P)       : 4
    Number of 1D Quadrature Points (Q) : 5
    Global nodes                       : 2560
    Local Elements                     : 75
    Owned nodes                        : 2560
    DoF per node                       : 3
  KSP:
    KSP Type                           : cg
    KSP Convergence                    : CONVERGED_RTOL
    Total KSP Iterations               : 35
    Final rnorm                        : 3.718027e-10
  Performance:
    Pointwise Error (max)              : 3.372489e-08
    CG Solve Time                      : 0.0147247 (0.0147247) sec
    DoFs/Sec in CG                     : 18.2551 (18.2551) million
```

`local_nodes 4000`
```
  Mesh:
    Number of 1D Basis Nodes (P)       : 4
    Number of 1D Quadrature Points (Q) : 5
    Global nodes                       : 4864
    Local Elements                     : 150
    Owned nodes                        : 4864
    DoF per node                       : 3
  KSP:
    KSP Type                           : cg
    KSP Convergence                    : CONVERGED_RTOL
    Total KSP Iterations               : 43
    Final rnorm                        : 3.267344e-10
  Performance:
    Pointwise Error (max)              : 5.538990e-08
    CG Solve Time                      : 0.0278914 (0.0278914) sec
    DoFs/Sec in CG                     : 22.4964 (22.4964) million
```

What am I missing?

File: examples/petsc/multigrid.c [hunk 1]
```diff
@@ -216,7 +216,7 @@ int main(int argc, char **argv) {
     ierr = DMSetVecType(dm[i], vec_type); CHKERRQ(ierr);
     PetscInt dim;
     ierr = DMGetDimension(dm[i], &dim); CHKERRQ(ierr);
-    ierr = SetupDMByDegree(dm[i], level_degrees[i], num_comp_u, dim,
+    ierr = SetupDMByDegree(dm[i], level_degrees[i], q_extra, num_comp_u, dim,
```
Comment (jeremylt) [inline examples/petsc/multigrid.c:219]: How are you keeping track of the fine level quadrature order in here?

Comment (rezgarshakeri) [reply examples/petsc/multigrid.c:219]: Do we use multigrid for `bp2`?

Comment (jeremylt) [reply examples/petsc/multigrid.c:219]: Not in `./bps`,but in `./multigrid` we certainly can.

I just don't see how the quadrature orders would agree here.

Comment (rezgarshakeri) [reply examples/petsc/multigrid.c:219]: Yeah, I didn't touch multigrid in this PR, in the older one I modified this line.
I wanted to narrow down the issue, so in this PR we only have new `basis_x` and `basis_u` (compare with main) and I cannot solve a simple projection problem.

Comment (rezgarshakeri) [reply examples/petsc/multigrid.c:219]: Hmm, this is basis_x in this PR
```
CeedBasis (H^1 space on a hexahedron element): dim=3 P=2 Q=3
      qref1d:    -0.77459667      0.00000000      0.77459667
   qweight1d:     0.55555556      0.88888889      0.55555556
    interp1d[0]:          0.88729833      0.11270167
    interp1d[1]:          0.50000000      0.50000000
    interp1d[2]:          0.11270167      0.88729833
      grad1d[0]:         -0.50000000      0.50000000
      grad1d[1]:         -0.50000000      0.50000000
      grad1d[2]:         -0.50000000      0.50000000
```

in main
```
CeedBasis (H^1 space on a hexahedron element): dim=3 P=2 Q=5
      qref1d:    -0.90617985     -0.53846931      0.00000000      0.53846931      0.90617985
   qweight1d:     0.23692689      0.47862867      0.56888889      0.47862867      0.23692689
    interp1d[0]:          0.95308992      0.04691008
    interp1d[1]:          0.76923466      0.23076534
    interp1d[2]:          0.50000000      0.50000000
    interp1d[3]:          0.23076534      0.76923466
    interp1d[4]:          0.04691008      0.95308992
      grad1d[0]:         -0.50000000      0.50000000
      grad1d[1]:         -0.50000000      0.50000000
      grad1d[2]:         -0.50000000      0.50000000
      grad1d[3]:         -0.50000000      0.50000000
      grad1d[4]:         -0.50000000      0.50000000
```

Comment (jeremylt) [reply examples/petsc/multigrid.c:219]: Oh, sometimes PETSc mangles the initial value of a option if the option is not found

File: examples/petsc/src/petscutils.c [hunk 2]
```diff
@@ -167,3 +170,209 @@ PetscErrorCode CreateRestrictionFromPlex(Ceed ceed, DM dm, CeedInt height,
...
+    PetscCall(PetscFEGetSpatialDimension(fe, &dim));
+    PetscCall(PetscFEGetNumComponents(fe, &num_comp));
+    PetscCall(PetscFEGetDualSpace(fe, &dual_space));
+    PetscCall(PetscDualSpaceGetDimension(dual_space, &num_dual_basis_vectors));
+    P = num_dual_basis_vectors / num_comp;
+    PetscCall(PetscQuadratureGetData(quadrature, NULL, NULL, &Q, NULL, NULL));
+
+    CeedInt P_1d = (CeedInt)round(pow(P, 1.0 / dim));
+    CeedInt Q_1d = (CeedInt)round(pow(Q, 1.0 / dim));
+
```
Comment (rezgarshakeri) [inline examples/petsc/src/petscutils.c:370]: Here returns Q_1d = 3 which makes sense for `p_degree=3`.
But in main we have `P= p_degree +1`, and `Q=P + q_extra` which gives you 5.

File: examples/petsc/multigrid.c (hunk 1 above)
Comment (jeremylt) [reply examples/petsc/multigrid.c:219]: If I manually set `-q_extra 1`, I seem to get the same results as main

Comment (jeremylt) [reply examples/petsc/multigrid.c:219]:
```diff
$ git diff
diff --git a/examples/petsc/bps.c b/examples/petsc/bps.c
index c97cc9e1..8e9b9f4c 100644
--- a/examples/petsc/bps.c
+++ b/examples/petsc/bps.c
@@ -414,7 +414,7 @@ static PetscErrorCode Run(RunParams rp, PetscInt num_resources,
     // Create DM
     PetscInt dim;
     ierr = DMGetDimension(dm_deg, &dim); CHKERRQ(ierr);
-    ierr = SetupDMByDegree(dm_deg, rp->degree, q_extra, rp->num_comp_u, dim,
+    ierr = SetupDMByDegree(dm_deg, rp->degree, rp->q_extra, rp->num_comp_u, dim,
                            bp_options[rp->bp_choice].enforce_bc,
                            bp_options[rp->bp_choice].bc_func); CHKERRQ(ierr);
     for (PetscInt r = 0; r < num_resources; r++) {
```

Comment (rezgarshakeri) [reply examples/petsc/multigrid.c:219]: That's weird, q_extra is defined few lines above and it is equal to `rp->q_extra`

File: examples/petsc/src/petscutils.c (hunk 2 above)
Comment (jeremylt) [reply examples/petsc/src/petscutils.c:370]: See diff above - small bug in older logic

**Pull request closed without merging**